File: python/make_cal.py
#!/usr/bin/env python
# coding: utf-8

# In[1]:


def is_notebook() -> bool:
    try:
        shell = get_ipython().__class__.__name__
        if shell == 'ZMQInteractiveShell':
            return True   # Jupyter notebook or qtconsole
        elif shell == 'TerminalInteractiveShell':
            return False  # Terminal running IPython
        else:
            return False  # Other type (?)
    except NameError:
        return False      # Probably standard Python interpreter

if is_notebook():
    get_ipython().run_line_magic('load_ext', 'autoreload')
    get_ipython().run_line_magic('autoreload', '2')

import argparse
import math
import os
from datetime import date
from svglib.svglib import svg2rlg
from reportlab.graphics import renderPDF

try:
    from IPython.display import SVG
except ImportError:
    SVG = None

# custom libs
from primitives import *
from arc_drawing import *
from calendar_data import *
from calendar_drawings import *
from layout import (
    calculate_layout_params,
    date_rotation,
    offset_point_by as offsetPointBy,
    build_solar_month_instances,
    build_islamic_month_instances,
    calculate_circle_rotation,
)
import islamic_alignment
import pdfizer


def svg_to_pdf(svg_path: str, pdf_path: str):
    """Convert SVG file to PDF using svglib and reportlab (pure Python)."""
    drawing = svg2rlg(svg_path)
    renderPDF.drawToFile(drawing, pdf_path)


def parse_args():
    """Parse command-line arguments for date specification."""
    parser = argparse.ArgumentParser(
        description="Generate circular calendar with Islamic/Gregorian overlay"
    )
    parser.add_argument(
        "--date",
        type=str,
        help="Gregorian date to align to (YYYY-MM-DD format). Defaults to today.",
    )
    parser.add_argument(
        "--hijri",
        type=str,
        help="Hijri date to align to (YYYY-MM-DD format, e.g., 1447-08-17 for Sha'ban 17, 1447).",
    )
    return parser.parse_args() if not is_notebook() else argparse.Namespace(date=None, hijri=None)


def get_alignment_from_args(args) -> islamic_alignment.AlignmentParams:
    """Get alignment parameters based on CLI arguments."""
    if args.hijri:
        parts = args.hijri.split("-")
        if len(parts) != 3:
            raise ValueError("Hijri date must be in YYYY-MM-DD format")
        hijri_year, hijri_month, hijri_day = int(parts[0]), int(parts[1]), int(parts[2])
        return islamic_alignment.get_alignment_params(
            hijri_year=hijri_year, hijri_month=hijri_month, hijri_day=hijri_day
        )
    elif args.date:
        parts = args.date.split("-")
        if len(parts) != 3:
            raise ValueError("Date must be in YYYY-MM-DD format")
        gregorian_date = date(int(parts[0]), int(parts[1]), int(parts[2]))
        return islamic_alignment.get_alignment_params(gregorian_date=gregorian_date)
    else:
        return islamic_alignment.get_alignment_params()


def main():
    # Parse arguments and calculate alignment
    args = parse_args()
    alignment = get_alignment_from_args(args)
    islamic_alignment.print_alignment_info(alignment)

    # Rotate Islamic months to start with current month
    islamic_year = Year(
        months=islamic_alignment.rotate_months(
            islamic_year_canonical.months, alignment.current_month_index
        )
    )

    # Derived alignment values
    islamic_date_rotation_offset = alignment.rotation_offset
    days_elapsed_islamic_base = alignment.days_elapsed


    # In[2]:


    # make the out dir
    os.makedirs("out", exist_ok=True)


    # In[3]:


    # Calculate layout
    layout = calculate_layout_params()
    scale_factor = layout.scale_factor

    print(scale_factor)

    solarMonths = build_solar_month_instances(
        solar_year, layout.date_box_height, layout.inner_radius, layout.outermost_radius
    )

    islamicMonths = build_islamic_month_instances(
        islamic_year, layout.date_box_height, layout.inner_radius,
        layout.outermost_radius, layout.month_thickness, islamic_date_rotation_offset
    )


    # In[4]:


    origin_first = layout.origin

    days_in_year = 366

    NUM_ROWS = layout.num_rows
    NUM_COLUMNS = layout.num_columns

    month_idx = 0
    page_num = 0
    pdfs = []

    # Draw full scale pages
    while month_idx < len(solarMonths) - 1:
        dwg = getVerticalPageCanvas() # getPageCanvas()
        origin = origin_first
        for col_num in range(NUM_COLUMNS):
            for row_num in range(NUM_ROWS):
                month_idx = row_num + NUM_ROWS*col_num + NUM_ROWS * NUM_COLUMNS * page_num

                if month_idx < len(solarMonths):
                    drawMonthParts(dwg, getMonth(solarMonths[month_idx], days_in_year, origin))
                    origin = offsetPointBy(origin, 0, layout.month_offset)

                if month_idx < len(islamicMonths):
                    drawMonthParts(dwg, getMonth(islamicMonths[month_idx], days_in_year, origin))
                    origin = offsetPointBy(origin, 0, (layout.month_thickness + layout.month_offset)*2.3)

            # move to next column
            origin = offsetPointBy(origin_first, layout.width * 1.05, 0)

        svg_file = f"out/svg_{scale_factor}_{page_num}.svg"
        pdf_file = f"out/calendar_page_{scale_factor}_{page_num}.pdf"

        dwg.saveas(svg_file, pretty=True)
        svg_to_pdf(svg_file, pdf_file)
        os.remove(svg_file)
        pdfs.append(pdf_file)

        page_num += 1



    # Draw summary page
    def circle_form_factor(g, month, days_elapsed, days_in_year, origin):
        g.translate(75, 50)
        g.scale(0.3)
        rot = calculate_circle_rotation(days_elapsed, month.num_days, days_in_year)
        g.rotate(rot, center=origin)
        g.translate(0, origin_first.y - origin.y)

    month_idx = 0
    page_num = 0
    days_elapsed = 0 - solarMonths[0].num_days/2

    # days_elapsed_islamic is now calculated dynamically from alignment params
    # The base value comes from islamic_alignment module; we just adjust for centering
    days_elapsed_islamic = days_elapsed_islamic_base - islamicMonths[0].num_days/2

    dwg = getVerticalPageCanvas() # getPageCanvas()
    origin_first = Point(layout.width_center, layout.outermost_radius + layout.vertical_offset)

    while month_idx < len(solarMonths) - 1:
        origin = Point(origin_first.x, origin_first.y)
        for col_num in range(NUM_COLUMNS):
            for row_num in range(NUM_ROWS):
                month_idx = row_num + NUM_ROWS*col_num + NUM_ROWS * NUM_COLUMNS * page_num

                if month_idx < len(solarMonths):
                    month = solarMonths[month_idx]
                    g = groupWithMonthParts(getMonth(month, days_in_year, origin))
                    circle_form_factor(g, month, days_elapsed, days_in_year, origin)
                    days_elapsed += month.num_days
                    dwg.add(g)
                    #origin = offsetPointBy(origin, 0, month_offset)


                if month_idx < len(islamicMonths):
                    month = islamicMonths[month_idx]
                    g = groupWithMonthParts(getMonth(month, days_in_year, origin))
                    circle_form_factor(g, month, days_elapsed_islamic, days_in_year, origin)
                    days_elapsed_islamic += month.num_days
                    dwg.add(g)
                    #origin = offsetPointBy(origin, 0, month_thickness*2.3)

            # move to next column
            #origin = offsetPointBy(origin_first, width * 1.05, 0)

        page_num += 1

    print("saving pdf")
    svg_file = f"out/svg_{scale_factor}_{page_num}_cover.svg"
    pdf_file = f"out/calendar_page_{scale_factor}_{page_num}_cover.pdf"

    dwg.saveas(svg_file, pretty=True)
    svg_to_pdf(svg_file, pdf_file)
    os.remove(svg_file)
    pdfs.insert(0, pdf_file)

    # Generate the instructions PDF with the cover image embedded
    import generate_instructions
    cover_pdf = f"out/calendar_page_{scale_factor}_{page_num}_cover.pdf"
    instructions_pdf = "v3 Instructions.pdf"
    generate_instructions.generate_instructions_pdf(cover_pdf, instructions_pdf)

    # Remove cover from pdfs list since it's now embedded in instructions
    pdfs.remove(cover_pdf)

    date_suffix = alignment.gregorian_date.strftime("%Y-%m-%d")
    pdfizer.concat_pdfs([instructions_pdf] + pdfs, f"out/calendar_pages_{scale_factor}_COMPLETE_{date_suffix}.pdf")
    print("Wrote the concatenated file!")

    # Clean up intermediate PDFs
    for pdf in pdfs:
        print(f"removing {pdf}...")
        os.remove(pdf)
    print(f"removing {cover_pdf}...")
    os.remove(cover_pdf)
    print("and removed old pdfs")

    print(scale_factor)
    if SVG:
        SVG(dwg.tostring())


if __name__ == "__main__" or is_notebook():
    main()

File: node/src/make-cal.ts
#!/usr/bin/env node
/**
 * Main entry point for circular calendar generation.
 * Ported from Python make_cal.py
 */

import { Command } from 'commander';
import * as fs from 'fs/promises';
import * as path from 'path';

import { SCALE_FACTOR, Point, resetTextPathIdCounter } from './primitives.js';
import { inchToMillimeter, getVerticalPageCanvas, drawMonthParts, groupWithMonthParts } from './arc-drawing.js';
import { getMonth } from './calendar-drawings.js';
import {
  type MonthInstance,
  type Month,
  solarYear,
  islamicYearCanonical,
} from './calendar-data.js';
import {
  getAlignmentParams,
  rotateMonths,
  printAlignmentInfo,
  type AlignmentParams,
} from './islamic-alignment.js';
import { svgToPng } from './svg-to-png.js';
import { createPdfFromPng, concatPdfs } from './pdfizer.js';
import { generateInstructionsPdf } from './generate-instructions.js';

// ============================================================
// CLI Argument Parsing
// ============================================================

export interface ParsedArgs {
  date?: string;
  hijri?: string;
}

export function parseArgs(args: string[]): ParsedArgs {
  const program = new Command();
  program
    .option('--date <date>', 'Gregorian date (YYYY-MM-DD)')
    .option('--hijri <date>', 'Hijri date (YYYY-MM-DD)')
    .parse(['node', 'make-cal', ...args]);

  return program.opts();
}

export function getAlignmentFromArgs(args: ParsedArgs): AlignmentParams {
  if (args.hijri) {
    const parts = args.hijri.split('-');
    if (parts.length !== 3) {
      throw new Error('Hijri date must be in YYYY-MM-DD format');
    }
    return getAlignmentParams(undefined, {
      hijriYear: parseInt(parts[0]),
      hijriMonth: parseInt(parts[1]),
      hijriDay: parseInt(parts[2]),
    });
  } else if (args.date) {
    const parts = args.date.split('-');
    if (parts.length !== 3) {
      throw new Error('Date must be in YYYY-MM-DD format');
    }
    return getAlignmentParams(
      new Date(parseInt(parts[0]), parseInt(parts[1]) - 1, parseInt(parts[2]))
    );
  }
  return getAlignmentParams();
}

// ============================================================
// Layout Configuration
// ============================================================

const scaleFactor = SCALE_FACTOR;
const canvasWidth = 11 * scaleFactor;
const width = inchToMillimeter(8 * scaleFactor);
const outermostRadius = width / (2 * 3.14 / 12);
const innerRadius = outermostRadius * 9.2 / 10;
const monthThickness = outermostRadius - innerRadius;
const dateBoxHeight = monthThickness * 0.2;

const extraWidthOffset = 10;
const widthCenter = inchToMillimeter(canvasWidth) / 2 + extraWidthOffset;
const verticalOffset = 30 * scaleFactor;

// Layout configuration based on scale factor
const NUM_ROWS = scaleFactor <= 0.5 ? 5 : scaleFactor < 0.75 ? 4 : 2;
const NUM_COLUMNS = scaleFactor <= 0.5 ? 2 : 1;

const DAYS_IN_YEAR = 366;

// ============================================================
// Month Instance Creation
// ============================================================

/**
 * Calculate date rotation offset for a month.
 */
export function dateRotation(month: Month): number {
  return -1 * (month.number - 1) * (360 / 12);
}

/**
 * Create MonthInstance array for solar (Gregorian) months.
 */
export function createSolarMonthInstances(): MonthInstance[] {
  const solarMonths: MonthInstance[] = [];

  for (let index = 0; index < solarYear.months.length; index++) {
    const month = solarYear.months[index];
    const nameUpsideDown = index >= 3 && index < 9;

    solarMonths.push({
      name: month.name,
      numDays: month.numDays[0],
      color: month.color,
      nameUpsideDown,
      dateOnTop: false, // Outer month
      dateBoxHeight,
      innerRadius,
      outerRadius: outermostRadius,
      dateAngleOffset: dateRotation(month),
    });
  }

  return solarMonths;
}

/**
 * Create MonthInstance array for Islamic months.
 */
export function createIslamicMonthInstances(alignment: AlignmentParams): MonthInstance[] {
  // Rotate Islamic months to start with current month
  const rotatedMonths = rotateMonths(
    islamicYearCanonical.months,
    alignment.currentMonthIndex
  );

  const islamicMonths: MonthInstance[] = [];
  const islamicRotationOffset = alignment.rotationOffset;

  for (const month of rotatedMonths) {
    const nameUpsideDown = month.number > 3 && month.number <= 9;

    islamicMonths.push({
      name: month.name,
      numDays: month.numDays[0],
      color: month.color,
      nameUpsideDown,
      dateOnTop: true, // Inner month
      dateBoxHeight,
      innerRadius: innerRadius - monthThickness,
      outerRadius: outermostRadius - monthThickness,
      dateAngleOffset: dateRotation(month) + islamicRotationOffset,
    });
  }

  return islamicMonths;
}

// ============================================================
// SVG Generation
// ============================================================

/**
 * Generate SVG for a calendar page (linear strips).
 */
export function generateCalendarSvg(
  solarMonths: MonthInstance[],
  islamicMonths: MonthInstance[]
): string {
  resetTextPathIdCounter();

  const monthOffset = 4 * scaleFactor;
  const originFirst = new Point(widthCenter, outermostRadius + verticalOffset);

  const svgElements: string[] = [];

  // SVG header
  const pageWidth = inchToMillimeter(8.5);
  const pageHeight = inchToMillimeter(11);
  svgElements.push(
    `<svg xmlns="http://www.w3.org/2000/svg" width="${pageWidth}mm" height="${pageHeight}mm" viewBox="0 0 ${pageWidth} ${pageHeight}">`
  );

  let origin = new Point(originFirst.x, originFirst.y);

  for (let i = 0; i < solarMonths.length; i++) {
    // Draw solar month
    const solarParts = getMonth(solarMonths[i], DAYS_IN_YEAR, origin);
    svgElements.push(drawMonthParts(solarParts));

    origin = new Point(origin.x, origin.y + monthOffset);

    // Draw Islamic month
    if (i < islamicMonths.length) {
      const islamicParts = getMonth(islamicMonths[i], DAYS_IN_YEAR, origin);
      svgElements.push(drawMonthParts(islamicParts));
    }

    origin = new Point(origin.x, origin.y + (monthThickness + monthOffset) * 2.3);
  }

  svgElements.push('</svg>');
  return svgElements.join('\n');
}

/**
 * Apply circular transform to a group of month parts.
 */
function circleFormFactor(
  svgContent: string,
  month: MonthInstance,
  daysElapsed: number,
  origin: Point,
  originFirst: Point
): string {
  const rotation = (360.0 * (daysElapsed + month.numDays / 2)) / DAYS_IN_YEAR;
  const translateY = originFirst.y - origin.y;

  return `<g transform="translate(75, 50) scale(0.3) rotate(${rotation}, ${origin.x}, ${origin.y}) translate(0, ${translateY})">
${svgContent}
</g>`;
}

/**
 * Generate SVG for the circular cover page.
 */
export function generateCircleCoverSvg(
  solarMonths: MonthInstance[],
  islamicMonths: MonthInstance[],
  daysElapsedIslamicBase: number
): string {
  resetTextPathIdCounter();

  const originFirst = new Point(widthCenter, outermostRadius + verticalOffset);
  const svgElements: string[] = [];

  // SVG header
  const pageWidth = inchToMillimeter(8.5);
  const pageHeight = inchToMillimeter(11);
  svgElements.push(
    `<svg xmlns="http://www.w3.org/2000/svg" width="${pageWidth}mm" height="${pageHeight}mm" viewBox="0 0 ${pageWidth} ${pageHeight}">`
  );

  let daysElapsed = -solarMonths[0].numDays / 2;
  let daysElapsedIslamic = daysElapsedIslamicBase - islamicMonths[0].numDays / 2;

  const origin = new Point(originFirst.x, originFirst.y);

  for (let i = 0; i < solarMonths.length; i++) {
    // Draw solar month in circular form
    const solarParts = getMonth(solarMonths[i], DAYS_IN_YEAR, origin);
    const solarSvg = drawMonthParts(solarParts);
    svgElements.push(
      circleFormFactor(solarSvg, solarMonths[i], daysElapsed, origin, originFirst)
    );
    daysElapsed += solarMonths[i].numDays;

    // Draw Islamic month in circular form
    if (i < islamicMonths.length) {
      const islamicParts = getMonth(islamicMonths[i], DAYS_IN_YEAR, origin);
      const islamicSvg = drawMonthParts(islamicParts);
      svgElements.push(
        circleFormFactor(islamicSvg, islamicMonths[i], daysElapsedIslamic, origin, originFirst)
      );
      daysElapsedIslamic += islamicMonths[i].numDays;
    }
  }

  svgElements.push('</svg>');
  return svgElements.join('\n');
}

// ============================================================
// Main Entry Point
// ============================================================

async function svgToPdf(svgContent: string, pdfPath: string): Promise<void> {
  const result = await svgToPng(svgContent);
  await createPdfFromPng(result.pngBuffer, pdfPath, result.widthPts, result.heightPts);
}

async function main() {
  // Parse arguments
  const args = parseArgs(process.argv.slice(2));
  const alignment = getAlignmentFromArgs(args);
  printAlignmentInfo(alignment);

  // Create output directory
  await fs.mkdir('out', { recursive: true });

  // Create month instances
  const solarMonths = createSolarMonthInstances();
  const islamicMonths = createIslamicMonthInstances(alignment);

  console.log(`Scale factor: ${scaleFactor}`);

  // Generate calendar pages
  const pdfs: string[] = [];
  let pageNum = 0;
  let monthIdx = 0;

  while (monthIdx < solarMonths.length - 1) {
    const startIdx = monthIdx;
    const endIdx = Math.min(monthIdx + NUM_ROWS * NUM_COLUMNS, solarMonths.length);

    const svg = generateCalendarSvg(
      solarMonths.slice(startIdx, endIdx),
      islamicMonths.slice(startIdx, endIdx)
    );

    const svgFile = `out/svg_${scaleFactor}_${pageNum}.svg`;
    const pdfFile = `out/calendar_page_${scaleFactor}_${pageNum}.pdf`;

    await fs.writeFile(svgFile, svg);
    await svgToPdf(svg, pdfFile);
    await fs.unlink(svgFile);

    pdfs.push(pdfFile);
    monthIdx = endIdx;
    pageNum++;
  }

  // Generate circular cover page
  console.log('Generating cover page...');
  const coverSvg = generateCircleCoverSvg(
    solarMonths,
    islamicMonths,
    alignment.daysElapsed
  );

  const coverSvgFile = `out/svg_${scaleFactor}_${pageNum}_cover.svg`;
  const coverPdfFile = `out/calendar_page_${scaleFactor}_${pageNum}_cover.pdf`;

  await fs.writeFile(coverSvgFile, coverSvg);
  await svgToPdf(coverSvg, coverPdfFile);
  await fs.unlink(coverSvgFile);

  // Generate instructions PDF with cover image
  console.log('Generating instructions PDF...');
  const instructionsPdf = 'v3 Instructions.pdf';

  // Convert cover to PNG for instructions
  const coverPngResult = await svgToPng(coverSvg);
  await generateInstructionsPdf(coverPdfFile, instructionsPdf, coverPngResult.pngBuffer);

  // Concatenate all PDFs (instructions first, then calendar pages)
  const outputFile = `out/calendar_pages_${scaleFactor}_COMPLETE.pdf`;
  await concatPdfs([instructionsPdf, ...pdfs], outputFile);
  console.log(`Wrote the concatenated file: ${outputFile}`);

  // Clean up intermediate PDFs
  for (const pdf of pdfs) {
    console.log(`Removing ${pdf}...`);
    await fs.unlink(pdf);
  }
  console.log(`Removing ${coverPdfFile}...`);
  await fs.unlink(coverPdfFile);
  await fs.unlink(instructionsPdf);

  console.log('Done!');
  console.log(`Scale factor: ${scaleFactor}`);
}

// Run main if executed directly
const isMain = process.argv[1]?.endsWith('make-cal.ts') || process.argv[1]?.endsWith('make-cal.js');
if (isMain) {
  main().catch((err) => {
    console.error('Error:', err);
    process.exit(1);
  });
}
